docs/rfcs/0002_templates_mvp.md

```diff
@@ -0,0 +1,372 @@
+
+### DataTemplate
+
+Represents the "essence" of the template (like File does with Version). Also
+provides a container for different versions of the template, allowing us
+to capture the provenance of the templates.
+
+#### Database Table Schema
+
+| Name         | Type        | Description                                                                |
+| ------------ | ----------- | -------------------------------------------------------------------------- |
+| id           | varchar     | A human friendly identifier (Kids First style) for the submission template |
+| name         | varchar     | Short name for submission template                                         |
+| description  | text        | A short blurb describing the entity the submission template represents     |
+| created_at   | datetime    | Timestamp of creation                                                      |
+| modified_at  | datetime    | Timestamp of update                                                        |
+| creator      | foreign key | Foreign key to User that created the submission template                   |
+| organization | foreign key | Foreign key to Organization that owns the template package                 |
+
+#### Django Model Schema
+
+| Name                          | Type           | Description                                 |
+| ----------------------------- | -------------- | ------------------------------------------- |
+| All fields in DB table schema |
+| template_versions             | RelatedManager | List of template versions this template has |
+
+### TemplateVersion
+
+A `TemplateVersion` is a container for the tabular submission file
+(`submission_template_file`) and associated submission rules
+(`fields_schema`). See [Definition](#📝-definition) for more details.
+
+#### Database Table Schema
+
+| Name          | Type        | Description                                                                                                                                                                                        |
+| ------------- | ----------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| id            | varchar     | A human friendly identifier (Kids First style) for the submission template                                                                                                                         |
+| description   | text        | A short blurb describing this version of the template                                                                                                                                              |
+| fields_schema | JSON        | The template submission rules. A JSON blob with a standard structure which will encode the template field objects having things like: label, key, description, type, values_accepted, instructions |
```

Comment:
Is this expanded on somewhere? It seems like this is the functional bit of the template - is there a document that goes over the different rules? how they are read, how dynamic they are, if they can access row / column / table data etc.

Reply:
Hey @shanewilson, first, sorry for the long response! It might be worth it to have a chat about this later. I'm sure this response doesn't answer all your questions

So there isn't a separate document to describe the `fields_schema` at the moment. In this first iteration, the field definitions will be very simple (label/column name, key/canonical attribute, column type, and values accepted). Later on we may expand on this and allow field definitions to include some sort of validation functions (e.g. age range check). 

Maybe some background would be useful. Currently, most of the heavy lift data validation is done after the submitted data file has been mapped (manually done by an analyst) to a set of standard column names understood by the study creator's validation library. It checks for things like: participants with more than one gender value, specimens that are not linked to any participant, genomic files missing associated sequencing experiment information, etc.

The rough workflow todayish/near future is:

1. Analyst uploads a tabular file to Data Tracker which conforms to one of the Kids First data templates (See [this PR](https://github.com/kids-first/kf-api-study-creator/pull/571) for details)
2. Analyst creates a data review on Data Tracker
3. Analyst adds their files to the data review
4. Analyst pushes a button to run validation
5. Validation is able to run on any data files that conform to the Kids First templates

As you might imagine the current process ^ needs to change in order make the validation a bit less hard coded. This is where defining validators for each field might help us. This is pretty big topic though and will require a separate RFC. It might help if you [take a peek](https://kids-first.github.io/kf-lib-data-ingest/tutorial/testing.html#data-validation) at how our analysts use the Kids First ingest/validation library to prepare the data and validate. I'm sorry that documentation is so meager but it might help.

When you say `different rules` - are you talking about field validation?

Can you expand on what you mean by `how dynamic they are` and what you mean by `how they are read`? 

I wonder if it would help to show you the json schema / structure I'm thinking of right now for the `fields_schema`. I figured I would expand on this later on when I get close to implementing the template versions.

Reply:
Thanks for the response @znatty22 

> When you say different rules - are you talking about field validation?

yea I'm asking about built-in rules I guess. Like can the template say a column is OneOf("A","B","C") and is there a list of those built-in rules?

> Can you expand on what you mean by how dynamic they are and what you mean by how they are read?

by dynamic I mean can the rules reference other bits of data? Like instead of OneOf(A,B) it could be OneOf(column.fieldA.values) or whatever

by how are they read I'm really asking about text vs code - how does a rule get defined and how is that turned into code. Is the rule just text? and then there is some text parsing to figure out which validation function to run?
I know in the ICGC submission they evenually added in https://en.wikipedia.org/wiki/MVEL (I think) so rules could be fully dynamic. It was super terrible because you basically had to write code in a text box. Just wondering what the plan is for when we have to start handling more complicated and dynamic rules.

Reply:
@shanewilson Hopefully these [slides](https://docs.google.com/presentation/d/1Qb425xdqGt0VKlCO81edyCzlCqohm94J1QwxxGd2ajY/edit?usp=sharing) and [this write up](https://github.com/kids-first/kf-api-study-creator/blob/templates-rfc/docs/rfcs/0003_template_fields_mvp.md) will answer some of your questions!

Reply:
Relating to your last point on en/decoding validation logic - I haven't searched around much on the topic but I remember when I started doing this before, I had come across CUE - an open-source data validation language: https://cuelang.org/docs/about/. It looks kinda confusing tho lol. Anywho, I will think about this more deeply when we get closer to designing the next version of validation in Dewrangle